vpr/src/route/parallel_connection_router.cpp

```diff
@@ -295,6 +309,76 @@ static inline bool prune_node(RRNodeId inode,
     return false;
 }
 
+static inline bool should_not_explore_neighbors(RRNodeId inode,
+                                float new_total_cost,
+                                RRNodeId target_node,
+                                vtr::vector<RRNodeId, t_rr_node_route_inf>& rr_node_route_inf_,
+                                const t_conn_cost_params& params) {
+#ifdef IS_DETERMINISTIC
+    (void)target_node;
+    // For deterministic pruning, cannot enforce anything on the total cost since
+    // traversal order is not gaurenteed. However, since total cost is used as a
+    // "key" to signify that this node is the last node that was pushed, we can
+    // just check for equality. There is a chance this may cause some duplicates
+    // for the deterministic case, but thats ok they will be handled.
+    // TODO: Maybe consider having the non-deterministic version do this too.
+    if (new_total_cost != rr_node_route_inf_[inode].path_cost)
+        return true;
+    // Perform post-target pruning. If this is not done, there is a chance that
+    // several duplicates of a node is in the queue that will never reach the
+    // target better than what we found and they will explore all of their
+    // neighbors which is not good. This is done before obtaining the lock to
+    // prevent lock contention where possible.
+    if (inode != target_node) {
+        float new_back_cost = rr_node_route_inf_[inode].backward_path_cost;
+        float best_back_cost_to_target = rr_node_route_inf_[target_node].backward_path_cost;
+        if (deterministic_post_target_prune_node(new_total_cost, new_back_cost, best_back_cost_to_target, params))
+            return true;
+    }
+#else // IS_DETERMINISTIC
+    (void)params;
+    if (new_total_cost > rr_node_route_inf_[inode].path_cost)
+        return true;
+    if (inode != target_node && new_total_cost > rr_node_route_inf_[target_node].path_cost)
```

Comment:
You should also post_target_prune agains the back cost here. It should be safe to do the following:
```cpp
if (inode != target_node) {
  if (new_total_cost > rr_node_route_inf_[target_node].path_cost)
    return true;
  float new_back_cost = rr_node_route_inf_[inode].backward_path_cost;
  if (new_back_cost > rr_node_route_inf_[target_node].backward_path_cost)
    return true;
}
```

This is safe since if the best_backward path cost happened to change part-way through this method, this would still prune as expected.